### PR TITLE
[InstCombine] Prefer to keep power-of-2 constants when combining ashr exact and slt/ult of a constant

### PR DESCRIPTION
**Pull request by asb**
We have flexibility in what constant to use when combining an `ashr exact` with a slt or ult of a constant, and it's not possible to revisit this decision later in the compilation pipeline after the `ashr exact` is removed. Keeping a constant close to power-of-2 (pow2val + 1) should be no worse than neutral, and in some cases may allow better codegen later on for targets that can more cheaply generate power of 2 (which may be selectable if converting back to setle/setge) or near power of 2 constants.

Alive2 proofs:
<https://alive2.llvm.org/ce/z/2BmPnq> and
<https://alive2.llvm.org/ce/z/DtuhnR>

---
*Discussion*:

More test changes are obviously required if landing this, but I thought I'd put this out there to get feedback on the approach as I'm not sure there's any precedent on making a heuristic choice for constants in instcombine. If it were possible to do this later in the pipeline (e.g. riscvcodegenprepare or later) I would, but as noted above once you lose the `ashr exact` you no longer know it's legal to change the constant. I don't have a benchmark this makes a huge difference on, but came across it when chasing something else (and intuitively, it seems plausible that making this kind of compare materialisable with a single instruction may reduce register pressure in the right cases).

The motivating example was something like:
```
define dso_local signext i32 @foo(ptr noundef %0, ptr noundef %1) local_unnamed_addr #0 {
  %3 = ptrtoint ptr %1 to i64
  %4 = ptrtoint ptr %0 to i64
  %5 = sub i64 %3, %4
  %6 = sdiv exact i64 %5, 32
  %7 = icmp sle i64 %6, 64
  br i1 %7, label %8, label %9

8:                                                ; preds = %2
  call void @abort() #2
  unreachable

9:                                                ; preds = %2
  ret i32 0

}
```

You get this kind of code from pointer arithmetic like `end - beg <= CONST` in C.

With this patch this can be compiled to:
```
	sub	a1, a1, a0
	bseti	a0, zero, 11
	bge	a0, a1, .LBB0_2
...
```

But without it you get an awkward to generate constant:
```
	sub	a1, a1, a0
	lui	a0, 1
	addiw	a0, a0, -2017
	bge	a0, a1, .LBB0_2
...
```
I'm arguing that producing pow2 or close to pow2 constants if you can is a good default that shouldn't hurt and may help - but perhaps there are cases I haven't considered?

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2485]: Please add some negative tests.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2491]:
```suggestion
      // --> icmp slt/ult X, (C - 1) << ShAmtC) + 1
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2499]: Please drop this change.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2493]: Please provide the generalized proof. This simplification seems incorrect :(

Proof: https://alive2.llvm.org/ce/z/4oha7i

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (asb) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2493]: Thank you, let me take a closer look.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2486]: Doesn't `(C - 1).isPowerOf2()` preclude `C.isMinSignedValue()`?

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (asb) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2493]: Thank you for taking a look. I wasn't aware of the InstCombine contributors guide when first writing this patch (it was committed soon after) and it seems I fell into some traps it warned against.

I do however have generalised proofs now:
<https://alive2.llvm.org/ce/z/2BmPnq>
<https://alive2.llvm.org/ce/z/DtuhnR>

Note that the power of two precondition isn't added in those proofs as it's not relevant for correctness, just for whether we opt to do this transformation. The key precondition is ensure we don't overflow into the sign bit which I believe my implementation fails to do as it doesn't have an equivalent of ` if (ShiftedC.ashr(ShAmtVal) == C)`.

Could you please check you're happy those proofs demonstrate generalized correctness (assuming a properly guarded transform)? Thanks.

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2493]: The proof looks fine to me.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (asb) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]: Yes, I've removed that now. Thanks.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (asb) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:2481]: Tests are now added.

**Comment by asb** [comment]
I think all comments are now addressed - tests are added also.

**Comment by dtcxzyw** [comment]
The implementation LGTM. However, I don't see any performance improvement/regression according to the IR diff/instruction count. Can you provide some performance data for this patch? I guess it requires the zbs extension.

**Comment by dtcxzyw** [comment]
> However, I don't see any performance improvement/regression according to the IR diff/instruction count.

I will re-test this patch later as I noticed llvm-diff doesn't handle flags/attrs :(

**Comment by asb** [comment]
> The implementation LGTM. However, I don't see any performance improvement/regression according to the IR diff/instruction count. Can you provide some performance data for this patch? I guess it requires the zbs extension.

Yes, I'd expect no real difference without zbs. I don't have performance figures for a full benchmark suite - I was analysing bedcov.c from https://github.com/attractivechaos/plb2 vs GCC and although it doesn't show up high on the profile, noted gcc managed to select `bseti` where we didn't. I thought this might be a simple missed isel pattern and of course ended up finding it was related to us choosing a harder to efficiently generate constant in InstCombine.

**Comment by dtcxzyw** [comment]
> > The implementation LGTM. However, I don't see any performance improvement/regression according to the IR diff/instruction count. Can you provide some performance data for this patch? I guess it requires the zbs extension.
> 
> Yes, I'd expect no real difference without zbs. I don't have performance figures for a full benchmark suite - I was analysing bedcov.c from https://github.com/attractivechaos/plb2 vs GCC and although it doesn't show up high on the profile, noted gcc managed to select `bseti` where we didn't. I thought this might be a simple missed isel pattern and of course ended up finding it was related to us choosing a harder to efficiently generate constant in InstCombine.

The IR diff basically looks fine to me. I am investigating the regression in pbrt-v4: https://github.com/dtcxzyw/llvm-opt-benchmark/pull/589/files#r1592080470

**Comment by dtcxzyw** [comment]
> > > The implementation LGTM. However, I don't see any performance improvement/regression according to the IR diff/instruction count. Can you provide some performance data for this patch? I guess it requires the zbs extension.
> > 
> > 
> > Yes, I'd expect no real difference without zbs. I don't have performance figures for a full benchmark suite - I was analysing bedcov.c from https://github.com/attractivechaos/plb2 vs GCC and although it doesn't show up high on the profile, noted gcc managed to select `bseti` where we didn't. I thought this might be a simple missed isel pattern and of course ended up finding it was related to us choosing a harder to efficiently generate constant in InstCombine.
> 
> The IR diff basically looks fine to me. I am investigating the regression in pbrt-v4: https://github.com/dtcxzyw/llvm-opt-benchmark/pull/589/files#r1592080470

Missing fold: https://alive2.llvm.org/ce/z/eUii9T
I will post a patch later to fix this.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for another approval.

**Review by nikic — APPROVED** [review]

**Pull request merged**
